src/Database/Expression/WindowExpression.php

```diff
@@ -83,22 +94,102 @@ public function order($fields)
      */
     public function range(?int $start, ?int $end = 0)
     {
-        return $this;
+        if (func_num_args() === 1) {
+            return $this->frame(self::RANGE, $start, self::PRECEDING);
+        }
+
+        return $this->frame(self::RANGE, $start, self::PRECEDING, $end, self::FOLLOWING);
     }
 
     /**
      * @inheritDoc
      */
     public function rows(?int $start, ?int $end = 0)
     {
-        return $this;
+        if (func_num_args() === 1) {
+            return $this->frame(self::ROWS, $start, self::PRECEDING);
+        }
+
+        return $this->frame(self::ROWS, $start, self::PRECEDING, $end, self::FOLLOWING);
     }
 
     /**
      * @inheritDoc
      */
     public function groups(?int $start, ?int $end = 0)
     {
+        if (func_num_args() === 1) {
+            return $this->frame(self::GROUPS, $start, self::PRECEDING);
+        }
+
+        return $this->frame(self::GROUPS, $start, self::PRECEDING, $end, self::FOLLOWING);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function frame(
+        int $type,
+        $startOffset,
+        int $startDirection,
+        $endOffset = null,
+        int $endDirection = self::FOLLOWING
+    ) {
```

Comment:
I have the impression that this should be an object builder instead of a function. It has many arguments of similar types and it will be difficult to remember the order when reading the code. Would it be possible to make this function return a builder or maybe accept a closure with a builder?

Reply:
Also even though there are constants for the directions the `int` typehint is disconcerting to read. Personally I would prefer the constants were string.

Reply:
How would you chain arguments if this returns a builder?

If we have another builder here, that would simply have 2 functions. One for start and one for end. Is that really necessary? Wouldn't it be confusing to have the simle range/rows/groups helper and then a builder?

Reply:
I think strings would be cleaner too. I went with constants based on feedback, but maybe constants as strings is a good way of doing both.

Reply:
Changed constants to strings and removed the verbose validation in frame().

I'm against a builder as it makes this very complex for two calls: start() and end(). 90% of users will just use the above helpers.